### PR TITLE
Subset mt to test

### PR DESCRIPTION
**Pull request by MattWellie**
Note: This is effectively a vastly improved non-dataproc version of [this fewgenomes script](https://github.com/populationgenomics/fewgenomes/blob/main/datasets/acute-care/extract_trio_vcf.py). The intended use case would be extracting one/all variants for some/all samples in a joint call, so that the calls can be interrogated either in Hail or using VCF tools when drilling into analysis of specific variants.

Simple script to take a target MT and produce a subset of it in test. Uses the output setting from the analysis-runner invocation to set the output path (Note: even when run in standard mode, the output path will be overwritten to point to test), and an `--out` argument to set a suffix within that folder for the MT/VCF. e.g. output will be:

`<dataset_bucket>-test/<output_path>/<suffix>.mt`

optional CLI arguments: 

    - a list of samples (`-s Sam1 Sam2 Sam3`): script checks that all requested samples are present (or fail), then subsets the MT columns to those samples only
    - a chrom & pos (--chr chrX --pos 12345): creates a Hail locus and filters to only this site
    - vcf flag (--vcf): creates a VCF of the requested data, and corresponding tabix index

if none of these arguments are specified, whole MT is copied with no VCF created

Discussion:

- if VCF flag is used, should we write a VCF instead of the MT (current behaviour is MT always, VCF _as well_ if flag is set)
- Commented out syntax would limit the MT to sites where the selected sample subset has a non-ref call. E.g. filtering the whole joint-call to one sample will contain all sites, but mainly consist of HomRef calls. The filter would reduce the MT/VCF to het/hom alt calls in the sub-selected sample(s)

### CODE REVIEW

File: data_transfer/subset_matrix_table.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
+#!/usr/bin/env python3
+
+"""
+Takes a path to a MatrixTable, and optionally sample IDs
+Reads in the MT from provided input path
+If sample IDs are specified, subset containing only those
+Write the [reduced] MT into Test
+Optionally with the --vcf flag, write as a VCF
+Optionally with both --chr and --pos specified, subset to a specific locus
```
Comment (cassimons) [inline data_transfer/subset_matrix_table.py]: Can we expand this to an interval (eg chr1:100-200) if we want more than a single base?

File: data_transfer/subset_matrix_table.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+    mt_samples = matrix.s.collect()
+
+    # check the subset exists
+    missing_samples = [sam for sam in samples if sam not in mt_samples]
+    if missing_samples:
+        raise Exception(f"Sample(s) missing from subset: {','.join(missing_samples)}")
+
+    filt_mt = matrix.filter_cols(hl.literal(set(samples)).contains(matrix.s))
+
+    # # optional - filter to variants with at least one alt call in these samples
```
Comment (cassimons) [inline data_transfer/subset_matrix_table.py]: ~~I don't think we want this by default.~~

I misunderstood. Yes I think this filter should be active by default.

**Comment by cassimons** [comment]
> if VCF flag is used, should we write a VCF instead of the MT (current behaviour is MT always, VCF as well if flag is set)

+0 on changing it it to a mt|vcf|both style option with mt being the default.

> Commented out syntax would limit the MT to sites where the selected sample subset has a non-ref call. E.g. filtering the whole joint-call to one sample will contain all sites, but mainly consist of HomRef calls. The filter would reduce the MT/VCF to het/hom alt calls in the sub-selected sample(s)

ahh I get what you mean now. Yes excluding sites where at least one selected sample is non-hom ref should be the default. However, we probably need an option to disable it. 

LGTM but would probably good to get Leo's blessing.

**Comment by MattWellie** [comment]
@cassimons added new commits to address both points 

- there is an `alts_only` flag which will do the 'hide all homref sites in this subset of samples'
- the --vcf argument is now --format mt|vcf|both

File: data_transfer/subset_matrix_table.py [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+
+    Returns
+    -------
+
+    """
+
+    mt_samples = matrix.s.collect()
+
+    # check the subset exists
+    missing_samples = [sam for sam in samples if sam not in mt_samples]
```
Comment (lgruen) [inline data_transfer/subset_matrix_table.py]:
```suggestion
    mt_samples = set(matrix.s.collect())

    # check the subset exists
    missing_samples = set(samples).difference(mt_samples)
```

File: data_transfer/subset_matrix_table.py [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    if missing_samples:
+        raise Exception(f"Sample(s) missing from subset: {','.join(missing_samples)}")
+
+    filt_mt = matrix.filter_cols(hl.literal(set(samples)).contains(matrix.s))
```
Comment (lgruen) [inline data_transfer/subset_matrix_table.py]:
```suggestion
    filt_mt = matrix.filter_cols(hl.set(samples).contains(matrix.s))
```

File: data_transfer/subset_matrix_table.py [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+import sys
+
+import hail as hl
+
+from cpg_utils.hail_batch import output_path, init_batch
+from cpg_utils.config import get_config
+
+
+def subset_to_samples(
+    matrix: hl.MatrixTable, samples: list[str], remove_hom_ref: bool
```
Comment (lgruen) [inline data_transfer/subset_matrix_table.py]: As a general style note, most Hail code I've seen just uses `mt` to denote a matrix table and typically reuses the same variable, e.g. `mt = mt.filter_cols(hl.set(samples).contains(mt.s))`.

File: data_transfer/subset_matrix_table.py [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+) -> hl.MatrixTable:
+    """
+    reduce the MatrixTable to a subset
```
Comment (lgruen) [inline data_transfer/subset_matrix_table.py]:
```suggestion
    filter the MatrixTable to a subset
```

File: data_transfer/subset_matrix_table.py [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+    Parameters
+    ----------
+    matrix :
+    samples :
+    remove_hom_ref :
+
+    Returns
+    -------
+
```
Comment (lgruen) [inline data_transfer/subset_matrix_table.py]: Either add descriptions or remove this part of the documentation completely?

File: data_transfer/subset_matrix_table.py [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    matrix :
+    chrom :
+    pos :
+
+    Returns
+    -------
+
+    """
+
+    locus = hl.Locus(contig=chrom, position=pos, reference_genome="GRCh38")
```
Comment (lgruen) [inline data_transfer/subset_matrix_table.py]: Here and below: we typically use single quotes for strings, not double quotes.

File: data_transfer/subset_matrix_table.py [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+    filt_mt = matrix.filter_cols(hl.literal(set(samples)).contains(matrix.s))
+
+    # optional - filter to variants with at least one alt call in these samples
+    if remove_hom_ref:
+        filt_mt = filt_mt.filter_entries(filt_mt.GT.is_non_ref())
+
+    return filt_mt
+
+
+def subset_to_locus(matrix: hl.MatrixTable, chrom: str, pos: int) -> hl.MatrixTable:
```
Comment (lgruen) [inline data_transfer/subset_matrix_table.py]: Would an interval be more useful in practice than a single locus?

File: data_transfer/subset_matrix_table.py [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+
+    # create the output path; make sure we're only ever writing to test
+    actual_output_path = output_path(output_root).replace(
+        f'cpg-{get_config()["workflow"]["dataset"]}-main',
+        f'cpg-{get_config()["workflow"]["dataset"]}-test',
+    )
+
+    if out_format in ["mt", "both"]:
+        # write the MT to a new output path
+        matrix.write(f"{actual_output_path}.mt", overwrite=True)
```
Comment (lgruen) [inline data_transfer/subset_matrix_table.py]: Note that overwriting will only work in `full` permissions mode, not `standard`. Might be worth not overwriting be default?

**Review by lgruen — APPROVED** [review]
This is great, thanks!

File: data_transfer/subset_matrix_table.py (hunk 10 above)
Comment (MattWellie) [reply data_transfer/subset_matrix_table.py]: As the output is always redirected to test no matter the permission level, won't overwriting be possible?

File: data_transfer/subset_matrix_table.py (hunk 8 above)
Comment (MattWellie) [reply data_transfer/subset_matrix_table.py]: Aye, I have a file watcher for Black, but it doesn't have the standard pre-commit settings... I'll resolve that

File: data_transfer/subset_matrix_table.py (hunk 9 above)
Comment (MattWellie) [reply data_transfer/subset_matrix_table.py]: I was having too many headaches getting it to run at all, so stretching to an interval was on my to-do list but not yet actioned. I'll revisit that!

**Review by cassimons — APPROVED** [review]

File: data_transfer/subset_matrix_table.py (hunk 10 above)
Comment (lgruen) [reply data_transfer/subset_matrix_table.py]: Oh, of course, you're absolutely right.

**Comment by MattWellie** [comment]
@cassimons  @lgruen  I've expanded the scope of this a bit... New added files handle the pre-commit hooks and github workflow management (copying from the python template repository which came after this was created)

@illusional's previously existing script has had some changes to please the linting workflow (mostly " -> ', one variable name change)

The main script being added has the following behaviour:
--format [mt/vcf/both] sets the output format to be written
-s sample sample2 ... sets the samples to retain from the joint-call
--chr chrN --pos 1234[-4567] sets a locus to filter for. This is either a point change or an interval, with some light validation applied (start<=end, parse-able as ints)
--out sets the output name prefix (write location defined using `output_path`)
--keep_ref defaults to False. If the flag is set, all sites will be kept, even if the remaining samples are only HomRef at those locations

File: data_transfer/subset_matrix_table.py (hunk 8 above)
Comment (MattWellie) [reply data_transfer/subset_matrix_table.py]: updated repo to include python-standard pre-commit hooks

File: data_transfer/subset_matrix_table.py (hunk 1 above)
Comment (MattWellie) [reply data_transfer/subset_matrix_table.py]: Done!

File: data_transfer/subset_matrix_table.py (hunk 2 above)
Comment (MattWellie) [reply data_transfer/subset_matrix_table.py]: Filter is active by default, and `--keep_ref` is the new flag to keep all sites

**Pull request merged**